zcash_client_backend/src/data_api.rs

```diff
@@ -0,0 +1,264 @@
+
+    fn block_height_extrema(&self) -> Result<Option<(BlockHeight, BlockHeight)>, Self::Error>;
+
+    fn get_target_and_anchor_heights(
+        &self,
+    ) -> Result<Option<(BlockHeight, BlockHeight)>, Self::Error> {
+        self.block_height_extrema().map(|heights| {
+            heights.map(|(min_height, max_height)| {
+                let target_height = max_height + 1;
+
+                // Select an anchor ANCHOR_OFFSET back from the target block,
+                // unless that would be before the earliest block we have.
+                let anchor_height = BlockHeight::from(cmp::max(
+                    u32::from(target_height).saturating_sub(ANCHOR_OFFSET),
+                    u32::from(min_height),
+                ));
+
+                (target_height, anchor_height)
+            })
+        })
+    }
+
+    fn get_block_hash(&self, block_height: BlockHeight) -> Result<Option<BlockHash>, Self::Error>;
+
+    fn get_max_height_hash(&self) -> Result<Option<(BlockHeight, BlockHash)>, Self::Error> {
+        self.block_height_extrema()
+            .and_then(|extrema_opt| {
+                extrema_opt
+                    .map(|(_, max_height)| {
+                        self.get_block_hash(max_height)
+                            .map(|hash_opt| hash_opt.map(move |hash| (max_height, hash)))
+                    })
+                    .transpose()
+            })
+            .map(|oo| oo.flatten())
+    }
+
+    fn get_tx_height(&self, txid: TxId) -> Result<Option<BlockHeight>, Self::Error>;
+
+    fn get_address<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        account: AccountId,
+    ) -> Result<Option<PaymentAddress>, Self::Error>;
+
+    fn get_extended_full_viewing_keys<P: consensus::Parameters>(
+        &self,
+        params: &P,
+    ) -> Result<Vec<ExtendedFullViewingKey>, Self::Error>;
+
+    fn is_valid_account_extfvk<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        account: AccountId,
+        extfvk: &ExtendedFullViewingKey,
+    ) -> Result<bool, Self::Error>;
+
+    fn get_balance(&self, account: AccountId) -> Result<Amount, Self::Error>;
+
+    fn get_verified_balance(
+        &self,
+        account: AccountId,
+        anchor_height: BlockHeight,
+    ) -> Result<Amount, Self::Error>;
+
+    fn get_received_memo_as_utf8(
+        &self,
+        id_note: Self::NoteRef,
+    ) -> Result<Option<String>, Self::Error>;
+
+    fn get_sent_memo_as_utf8(&self, id_note: Self::NoteRef) -> Result<Option<String>, Self::Error>;
+
+    fn get_commitment_tree(
+        &self,
+        block_height: BlockHeight,
+    ) -> Result<Option<CommitmentTree<Node>>, Self::Error>;
+
+    fn get_witnesses(
+        &self,
+        block_height: BlockHeight,
+    ) -> Result<Vec<(Self::NoteRef, IncrementalWitness<Node>)>, Self::Error>;
+
+    fn get_nullifiers(&self) -> Result<Vec<(Vec<u8>, AccountId)>, Self::Error>;
+
+    fn select_spendable_notes(
```

Comment:
Yeah, this is one of those situations where it's a method that really should be implemented in terms of more primitive operations, but where we want the backend provider to be able to override that default selection algorithm for reasons of efficiency or user preference. I expect that this will need to change in phase 2.

Reply:
If the API provides all spendable notes, that means all spendable notes have to be locked while the wallet's note selection runs, right? Whereas this API potentially allows finer-grained locking. (Although now that I think about it, I'm not even sure that is a good idea. But it is an important semantic difference.)